alf/pretrained_models/model_adapters.py

```diff
@@ -0,0 +1,278 @@
+        """Check if the adapter class can adapt a given module.
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def _decompose_weight_dims(cls, m: nn.Module) -> Tuple[int]:
+        """Return two (reshaped) dimensions ``(nrows, ncols)`` so that
+        ``nrows*ncols == np.prod(m.weight.shape)``.
+
+        The reinterpreted shape will be used to create LoRA weights ``self._wA``
+        and ``self._wB``.
+
+        Returns:
+            tuple: a pair of ints ``(nrows, ncols)``.
+        """
+        raise NotImplementedError()
+
+    def forward(self, input):
+        raise NotImplementedError()
+
+
+@alf.configurable
+class EmbeddingAdapter(LoRA):
+    """Adapter for embedding layers.
+    """
+
+    @classmethod
+    def can_adapt(cls, m: nn.Module) -> bool:
+        return isinstance(m, nn.Embedding)
+
+    @classmethod
+    def _decompose_weight_dims(cls, m: nn.Module):
+        return m.num_embeddings, m.embedding_dim
+
+    def forward(self, input):
+        m = self._m[0]
+        embedding_table = self._wB or self._wA
+        input = F.embedding(input, embedding_table, m.padding_idx, m.max_norm,
+                            m.norm_type, m.scale_grad_by_freq, m.sparse)
+        if self._wB is not None:
+            input = input @ self._wA
+        return input
+
+
+@alf.configurable
+class LinearAdapter(LoRA):
+    """Adapter for linear layers.
+    """
+
+    @classmethod
+    def can_adapt(cls, m: nn.Module) -> bool:
+        return isinstance(m, nn.Linear)
+
+    @classmethod
+    def _decompose_weight_dims(cls, m: nn.Module):
+        return m.out_features, m.in_features
+
+    def forward(self, input):
+        input = input @ self._wA.t()
+        if self._wB is not None:
+            input = input @ self._wB.t()
+        return input
+
+
+@alf.configurable
+class Conv2dAdapter(LoRA):
+    """Adapter for Conv2d layers.
+
+    The most natural way of a LoRA decomposition for Conv2d is
+
+    .. code-block:: python
+
+        (rank, kernel_size[0] * kernel_size[1] * in_channels) x (out_channels, rank)
+
+    However, since ``out_channels`` is usually small, this decomposition is not
+    low-rank actually and it won't save much memory.
+
+    We can first reinterpret the weight matrix as a shape of
+
+    .. code-block:: python
+
+        (out_channels * kernel_size[0], in_channels * kernel_size[1])
+
+    to make the in- and out-dimensions balanced, and then decompose.
+    """
+
+    @classmethod
+    def can_adapt(cls, m: nn.Module) -> bool:
+        return isinstance(m, nn.Conv2d)
+
+    @classmethod
+    def _decompose_weight_dims(cls, m: nn.Module):
+        kernel_size = m.kernel_size
+        if not isinstance(kernel_size, tuple):
+            kernel_size = (kernel_size, ) * 2
+        nrows = m.out_channels // m.groups * kernel_size[0]
+        ncols = m.in_channels * kernel_size[1]
+        return nrows, ncols
+
+    def forward(self, input):
+        m = self._m[0]
+        return F.conv2d(
```

Comment:
The following might be more efficient (need to figure `...` part):
```python
a=F.conv2d(input, wA.reshape(r, w.shape[0],  kerneal_size[0], 1), ...)
out=F.conv2d(a, wB.reshape(w.shape[0],  r, 1, kerneal_size[1]), ...)
```

Reply:
> The following might be more efficient (at least in terms of FLOPs) (need to figure out `...` part):
> 
> ```python
> a=F.conv2d(input, wA.reshape(r, w.shape[0],  kerneal_size[0], 1), ...)
> out=F.conv2d(a, wB.reshape(w.shape[0],  r, 1, kerneal_size[1]), ...)
> ```

It's not easy to convolve in this two-stage way in general. For example, if m.padding > 0, then we shouldn't pad twice otherwise it introduces extra padding effects. If we add the padding only to the first/second conv, then the other one should have kernel_size=1 in order to avoid input size decrease. In this case, the low-rank decomposition has a very small benefit.

Reply:
> The following might be more efficient (at least in terms of FLOPs) (need to figure out `...` part):
> 
> ```python
> a=F.conv2d(input, wA.reshape(r, w.shape[0],  kerneal_size[0], 1), ...)
> out=F.conv2d(a, wB.reshape(w.shape[0],  r, 1, kerneal_size[1]), ...)
> ```

I didn't see a general way for this two-stage convolution? For example, if m.groups > 1, then r must be divisible by m.groups to have the correct input-output mapping.

Reply:
Perhaps we can just do this for the case of groups=1. As convs with groups > 1 are not that expensive compared with groups==1.